# Homework 1

## Question 1

In [2]:
import pandas as pd
url = "squirrel_data.csv"
df = pd.read_csv(url)
df.isna().sum()

X                                                0
Y                                                0
Unique Squirrel ID                               0
Hectare                                          0
Shift                                            0
Date                                             0
Hectare Squirrel Number                          0
Age                                            121
Primary Fur Color                               55
Highlight Fur Color                           1086
Combination of Primary and Highlight Color       0
Color notes                                   2841
Location                                        64
Above Ground Sighter Measurement               114
Specific Location                             2547
Running                                          0
Chasing                                          0
Climbing                                         0
Eating                                           0
Foraging                       

## Question 2

### 1.

In [4]:
import pandas as pd

# Load the dataset
squirrel_data = pd.read_csv("squirrel_data.csv")

# Get the number of rows and columns
rows, columns = squirrel_data.shape

print(f"The dataset has {rows} rows and {columns} columns.")


The dataset has 3023 rows and 31 columns.


### 2.

In a dataset, the **observations** are the individual data points or instances of data, which in this case is every sighting of a squirrel in Central Park.

The **variables** are all the individual pieces of information recorded for each observation in the dataset, which in this case includes information such as "Age", "Primary Fur Color" and "Highlight Fur Color" among others.

## Question 3

In [5]:
squirrel_data.describe()

X            Y          Date  Hectare Squirrel Number
count  3023.000000  3023.000000  3.023000e+03              3023.000000
mean    -73.967184    40.780853  1.011949e+07                 4.123718
std       0.007726     0.010285  4.246671e+04                 3.096492
min     -73.981159    40.764911  1.006202e+07                 1.000000
25%     -73.973102    40.771676  1.008202e+07                 2.000000
50%     -73.968594    40.778166  1.012202e+07                 3.000000
75%     -73.960189    40.791219  1.014202e+07                 6.000000
max     -73.949722    40.800119  1.020202e+07                23.000000

## Question 4

In [6]:
import pandas as pd
url = "titanic.csv"
df = pd.read_csv(url)
df.shape

(891, 15)

In [7]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

df.shape returns (891, 15), meaning that there are 891 observations and 15 variables in the titanic dataset. However, when we execute df.describe(), we only see the analysis of 6 out of the 15 variables, which all happen to be numeric values. We can conclude that df.describe() only returns information for the numeric values, and not for the 9 other non-numeric variables.
Furthermore, the "count" data for each variable is 891 except for the age variable, which has a "count" value of 714. Since 891 is the number of observations, we can conclude that there is missing data for the ages of certain observations.

## Question 5

The difference between an "attribute" and a "method" is that an "attribute" is a property of an object, that holds information relating to the object and that can be easily retrieved. For example, in df.shape, "shape" is an "attribute" of the df object that holds information about the amount of rows and columns in df. On the other hand, a "method" is a function (set of instructions) that is called within the context of an object. "Methods" have access to the "attributes" of the object that is calling it, and they can modify the object itself.

## Question 6

**'count':** This is the amount of non-missing values in the column.<br>
**'mean':** This is the average of the values in the column.<br>
**'std':** This is a measure of the amount of variation in the values of the column.<br>
**'min':** This is the smallest value in the column.<br>
**'25%':** This is the 25th percentile value. It's the value that is larger than or equal to 25% of the values in the column.<br>
**'50%':** This is the 50th percentile value, also known as the median. It's the value that is larger than or equal to 50% of the values in the column.<br>
**'75%':** This is the 75th percentile value. It's the value that is larger than or equal to 75% of the values in the column.<br>
**'max' :** This is the largest value in the column.

## Summary of conversation up to this point with ChatGPT

Link to conversation: https://chatgpt.com/share/67130503-93e5-48e4-ad52-6775f3de7fed<br>
Summary of conversation:<br>
Certainly! Here’s a summary of our exchanges:

---

### **Summary of Interactions**

1. **Dataset Overview:**
   - **Topic:** You are working with the "2018 Central Squirrel Census" dataset saved in a file named `squirrel_data.csv`.
   - **Task:** You wanted to know how to determine the number of rows and columns in the dataset using pandas.

   **Response:** 
   - You can load the dataset using `pandas` and use the `.shape` attribute to get the number of rows and columns.
   - Example code:
     ```python
     import pandas as pd
     squirrel_data = pd.read_csv("squirrel_data.csv")
     rows, columns = squirrel_data.shape
     print(f"The dataset has {rows} rows and {columns} columns.")
     ```

2. **Explanation of Observations and Variables:**
   - **Topic:** You asked about the meaning of "observations" and "variables" in the context of the dataset.

   **Response:**
   - **Observations** are individual data points or records (each row in the dataset). Each row represents a sighting of a squirrel.
   - **Variables** are the different types of information recorded for each observation (each column in the dataset). Variables include details like the squirrel’s color, age, and behavior.

3. **Summarizing Dataset Columns:**
   - **Topic:** You inquired about how to provide simple summaries of the columns in the dataset.

   **Response:**
   - Several methods were suggested to summarize the dataset:
     - `describe()`: Provides descriptive statistics for numerical columns.
     - `info()`: Gives a summary of the DataFrame, including data types and non-null counts.
     - `value_counts()`: Shows the distribution of values for categorical columns.
     - `isnull().sum()`: Counts missing values in each column.
     - `head()`: Displays the first few rows of the dataset for a quick overview.

4. **Difference Between Attributes and Methods:**
   - **Topic:** You asked about the difference between attributes (which don’t end with `()`) and methods (which do end with `()`).

   **Response:**
   - **Attributes**: Store data or state information about an object and are accessed without parentheses (e.g., `df.shape`).
   - **Methods**: Perform actions on the object and are called with parentheses (e.g., `df.describe()`).

5. **Explanation of Summary Statistics:**
   - **Topic:** You asked for an explanation of the summary statistics provided by the `describe()` method, including 'count', 'mean', 'std', 'min', '25%', '50%', '75%', and 'max'.

   **Response:**
   - Each of these statistics was explained:
     - **Count**: Number of non-null values.
     - **Mean**: Average of the values.
     - **Std**: Standard deviation, a measure of spread.
     - **Min**: Minimum value.
     - **25% (Q1)**: First quartile, where 25% of the data lies below.
     - **50% (Median/Q2)**: Middle value, where 50% of the data lies below.
     - **75% (Q3)**: Third quartile, where 75% of the data lies below.
     - **Max**: Maximum value.

---

This summary encapsulates the key points of our discussion and can be submitted as a record of your interactions for your homework assignment.

## Beginning of new chat with GPT

## Question 7

### 1

An example of a scenario in which using df.dropna() might be preferred over using del df['col'] is if we're preparing a dataset for a machine learning model that doesn't handle missing values well. In this case, if the data in a column is important to the model, it would be better to use df.dropna() to only remove the rows which contain missing data.

### 2

An example of a scenario in which using del df['col'] might be preferred over using df.dropna() is if we're preparing a dataset for a machine learning model. We realize that the dataset has a whole lot of missing data, but only in 1 column that doesn't really interest us anyways for our model. In this case, using del df['col'] to delete that column entirely would be preferrable, so we can conserve the most observations possible after doing df.dropna().

### 3

Applying del df['col'] before df.dropna() when both are used together could be important, because if we're using a dataset in which there are certain columns, with missing data, that we aren't interested in for our use case, we could first delete those columns entirely using del df['col'], and then we can rid our dataset of observations that contain missing data in columns we are interested in using with df.dropna(). In the end, we may be able to conserve more rows in our cleaned dataset than if we hadn't done del df['col'] beforehand.

### 4

First off, I determine how much missing data is in the Titanic Dataset for each column:

In [16]:
df.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

I see that there is a lot of missing data in the "deck" column, a moderate amount of missing data in the "age" column relative to the total amount of observations, and very little missing data in the "embark_town" column. Since there is so much missing data in the "deck" column, I first delete that column entirely:

In [17]:
del df["deck"]

Then, since I believe that the "age" column contains interesting information on the passengers of the Titanic, I think it's a good idea to keep the column despite it making the dataset lose a minimum of 177 observations when executing df.dropna(). Finally, the "embark_town" column has such little missing data, it's worth keeping. I now execute df.dropna():

In [18]:
df = df.dropna()

In [19]:
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
885         0       3  female  39.0      0      5  29.1250        Q   Third   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male  embark_town alive  alone  
0      man        True  Southampton    no  False  
1    woman       False    Cherbourg   yes  False  
2    woman       False  Southampton   yes   True  
3    woman       False  Southampton   yes  False  
4      man        True  Southampton    no   True  
..     ...         ...          ...   ...    ...  
885  woman       False   Queenstown    no  False  
886    man        True  Southampton    no   True  
887  woman       False  Southampton   yes   True  
889    man        True    Cherbourg   yes   True  
890    man        True   Queenstown    no   True  

[712 rows x 14 columns]

Before, the uncleaned dataset had 891 rows and 15 columns. After, the cleaned dataset has 712 rows and 14 columns.

## Summary of conversation with ChatGPT

Link to conversation: https://chatgpt.com/share/cf64c1e2-779c-40a8-a6fa-0c90ff1cda5a<br>
Summary of conversation:<br>
Here’s a summary of our interactions:

1. **Use Case for `df.dropna()` vs. `del df['col']`:**
   - **Question:** You asked for an example of a use case where using `df.dropna()` might be preferred over using `del df['col']`.
   - **Initial Explanation:** I explained that `df.dropna()` is useful for removing rows with any missing values, while `del df['col']` is used to remove an entire column. The key difference is that `df.dropna()` cleans the DataFrame by removing rows with missing data, while `del df['col']` removes a specific column entirely.
   
2. **Tangible Scenario with Real Dataset:**
   - **Request:** You asked for a more tangible scenario based on a real dataset.
   - **Titanic Dataset Example:** I provided an example using the Titanic dataset, where `df.dropna()` could be used to remove rows with missing values in columns like `Age`, `Fare`, or `Cabin`, ensuring the dataset is clean before model training. I also explained that using `del df['Cabin']` would remove the entire `Cabin` column, potentially losing valuable information.

This summary reflects the main points of our discussion and can serve as a record for your homework assignment.

## Beginning of new chat with GPT

## Question 8

### 1

In [11]:
# reset dataset, with missing values
df = pd.read_csv(url)
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

The code `df.groupby("col1")["col2"].describe()` groups df by the values in the column "col1". Then, for each group obtained, we get the summary statistics for the values in the column "col2". For example, in the context of the Titanic Dataset, we can do:

In [32]:
df.groupby("pclass")["survived"].describe()

count      mean       std  min  25%  50%  75%  max
pclass                                                    
1       216.0  0.629630  0.484026  0.0  0.0  1.0  1.0  1.0
2       184.0  0.472826  0.500623  0.0  0.0  0.0  1.0  1.0
3       491.0  0.242363  0.428949  0.0  0.0  0.0  0.0  1.0

This code first groups each passenger in the dataset by their passenger class. Then, we get the summary statistics for the values in the "survived" column for each group. Thanks to this code, we can determine the survival rate for the passengers depending on their class:<br>
Class 1: 63%<br>
Class 2: 47%<br>
Class 3: 24%<br>

### 2

In [33]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [34]:
df.groupby("pclass")["age"].describe()

count       mean        std   min   25%   50%   75%   max
pclass                                                           
1       186.0  38.233441  14.802856  0.92  27.0  37.0  49.0  80.0
2       173.0  29.877630  14.001077  0.67  23.0  29.0  36.0  70.0
3       355.0  25.140620  12.495398  0.42  18.0  24.0  32.0  74.0

The values in count that you get from doing `df.groupby("col1")["col2"].describe()` are fundamentally different from the values in count you get from `df.describe()`, because while the count values you obtain by doing `df.describe()` only report the amount of non-missing values for each column, the count values you obtain by doing `df.groupby("col1")["col2"].describe()` tell you how many observations are in each group; in other words, for all the different values in the "col1" column, it'll tell you how many observations have that specific value for the "col1" variable.

## Summary of conversation with ChatGPT

Link to conversation: https://chatgpt.com/share/d1ae0791-e6bd-4835-8df0-7dbae7c29799<br>
Summary of conversation:<br>
Here’s a summary of our interactions:

1. **Understanding `df.groupby("col1")["col2"].describe()`:**
   - **Question:** You asked what `df.groupby("col1")["col2"].describe()` does in pandas.
   - **Explanation:** I explained that this code groups the DataFrame by the values in "col1" and calculates summary statistics (like count, mean, std, etc.) for "col2" within each group, helping to analyze the distribution of "col2" across different categories in "col1".

2. **Applying `groupby` to the Titanic Dataset:**
   - **Request:** You wanted to know how to use this code with the Titanic dataset.
   - **Examples Provided:** I gave examples, such as analyzing survival rates by passenger class, fare distribution by embarkation town, and age distribution by sex.

3. **How `groupby` Works:**
   - **Question:** You asked for clarification on how `groupby` performs grouping.
   - **Explanation:** I clarified that `groupby` organizes data into groups based on unique values in a specified column and doesn't sum other columns by default. Operations like summing or calculating means are applied after grouping.

4. **Interpreting the "mean" Column:**
   - **Question:** You inquired about what the "mean" column represents in the output of `df.groupby("pclass")["survived"].describe()`.
   - **Explanation:** I explained that the "mean" column shows the average survival rate within each passenger class, indicating the proportion of passengers who survived.

This summary encapsulates the main points of our discussion and can be used for your reference.

## Beginning of new chat with GPT

### 3

In [14]:
# code with errors in it
import pandas as pd
url = "titanic.csv"
df = pd.read_csv(url)
df.groupby("sex")["age"].describe()

count       mean        std   min   25%   50%   75%   max
sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0

**Error Debugging Report for ChatGPT**:<br>
Error A: Fixed it immediately<br>
Error B: ChatGPT didn't know to correct "titanics.csv" to "titanic.csv"<br>
Error C: Fixed it immediately<br>
Error D: Fixed it immediately<br>
Error E: Fixed it immediately<br>
Error F: Fixed it immediately<br>
Error G: Fixed it immediately<br>

**Error Debugging Report for Google**:<br>
Using Google to fix these, I do end up finding the solution most of the time, but it's more of a hassle and takes more time.<br><br>
In my opinion, it's easier to work with a ChatBot, as it found the solution to the problems most of the time and it did it very quickly.

## Summary of conversation with ChatGPT

Link to conversation: https://chatgpt.com/share/288c9817-8b41-45fc-bc4d-bf3bfc0e4d1d<br>
Summary of conversation:<br>
Certainly! Here's a summary of our interactions:

1. **Initial Issue with `pd` Not Defined**:
   - User encountered a `NameError: name 'pd' is not defined`.
   - Solution: Suggested importing `pandas` with `import pandas as pd` and ensuring `pandas` is installed.

2. **FileNotFoundError with `titanics.csv`**:
   - User faced a `FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'`.
   - Solution: Recommended checking the file path, ensuring the file name and extension are correct, and verifying that the file is in the correct directory.

3. **Issue with `NameError` for `DF`**:
   - User encountered a `NameError: name 'DF' is not defined`.
   - Solution: Pointed out the case sensitivity of variable names and suggested using the correct variable `df`.

4. **Syntax Error with Missing Parenthesis**:
   - User had a syntax error due to a missing closing parenthesis.
   - Solution: Corrected the syntax by adding the missing parenthesis.

5. **Error with `group_by` Method**:
   - User received an error due to using `group_by` instead of `groupby`.
   - Solution: Corrected the method name to `groupby`.

6. **KeyError for `'Sex'` Column**:
   - User received a `KeyError: 'Sex'`.
   - Solution: Suggested checking the column names, ensuring there are no typos or extra spaces, and using the correct column name.

7. **Fixing Column Name in GroupBy Operation**:
   - User had an issue with `df.groupby("sex")[age].describe()`.
   - Solution: Recommended enclosing `'age'` in quotes to correctly reference the column name.

If you need any more details or further clarification, just let me know!

## Question 9

Mostly.